File: Go/CatAndMouse.go
//go:build ignore

package main

import (
	"fmt"
	"slices"
)

const (
	DRAW  = 0
	MOUSE = 1
	CAT   = 2
)

func catMouseGame(graph [][]int) int {
	n := len(graph)
	color := make([][][3]int, n)
	degree := make([][][3]int, n)
	for i := range color {
		color[i] = make([][3]int, n)
		degree[i] = make([][3]int, n)
	}
	hasZero := make([]bool, n)
	for i, adj := range graph {
		hasZero[i] = slices.Contains(adj, 0)
	}
	for m := range n {
		for c := range n {
			degree[m][c][MOUSE] = len(graph[m])
			degree[m][c][CAT] = len(graph[c])
			if hasZero[c] {
				degree[m][c][CAT]--
			}
		}
	}

	type State struct{ m, c, turn, result int }
	queue := []State{}
	for i := range n {
		for _, t := range []int{MOUSE, CAT} {
			color[0][i][t] = MOUSE
			queue = append(queue, State{0, i, t, MOUSE})
			if i > 0 {
				color[i][i][t] = CAT
				queue = append(queue, State{i, i, t, CAT})
			}
		}
	}

	for len(queue) > 0 {
		s := queue[0]
		queue = queue[1:]
		if s.turn == MOUSE {
			for _, pc := range graph[s.c] {
				if pc == 0 || color[s.m][pc][CAT] != DRAW {
					continue
				}
				if s.result == CAT {
					color[s.m][pc][CAT] = CAT
					queue = append(queue, State{s.m, pc, CAT, CAT})
				} else {
					degree[s.m][pc][CAT]--
					if degree[s.m][pc][CAT] == 0 {
						color[s.m][pc][CAT] = s.result
						queue = append(queue, State{s.m, pc, CAT, s.result})
					}
				}
			}
		} else {
			for _, pm := range graph[s.m] {
				if color[pm][s.c][MOUSE] != DRAW {
					continue
				}
				if s.result == MOUSE {
					color[pm][s.c][MOUSE] = MOUSE
					queue = append(queue, State{pm, s.c, MOUSE, MOUSE})
				} else {
					degree[pm][s.c][MOUSE]--
					if degree[pm][s.c][MOUSE] == 0 {
						color[pm][s.c][MOUSE] = s.result
						queue = append(queue, State{pm, s.c, MOUSE, s.result})
					}
				}
			}
		}
	}

	return color[1][2][MOUSE]
}

func main() {
	fmt.Println(catMouseGame([][]int{{2, 5}, {3}, {0, 4, 5}, {1, 4, 5}, {2, 3}, {0, 2, 3}})) // 0
	fmt.Println(catMouseGame([][]int{{1, 3}, {0}, {3}, {0, 2}}))                              // 1
}

File: Rust/src/bin/cat_and_mouse.rs
use std::collections::VecDeque;

const DRAW: i32 = 0;
const MOUSE: i32 = 1;
const CAT: i32 = 2;

fn cat_mouse_game(graph: Vec<Vec<i32>>) -> i32 {
    let n = graph.len();
    let mut color = vec![vec![[DRAW; 3]; n]; n];
    let mut degree = vec![vec![[0i32; 3]; n]; n];
    let has_zero: Vec<bool> = graph.iter().map(|adj| adj.contains(&0)).collect();
    for m in 0..n {
        for c in 0..n {
            degree[m][c][MOUSE as usize] = graph[m].len() as i32;
            degree[m][c][CAT as usize] = graph[c].len() as i32 - if has_zero[c] { 1 } else { 0 };
        }
    }

    let mut queue: VecDeque<(usize, usize, i32, i32)> = VecDeque::new();
    for i in 0..n {
        for &t in &[MOUSE, CAT] {
            color[0][i][t as usize] = MOUSE;
            queue.push_back((0, i, t, MOUSE));
            if i > 0 {
                color[i][i][t as usize] = CAT;
                queue.push_back((i, i, t, CAT));
            }
        }
    }

    while let Some((m, c, turn, result)) = queue.pop_front() {
        if turn == MOUSE {
            for &pc_i in &graph[c] {
                let pc = pc_i as usize;
                if pc_i == 0 || color[m][pc][CAT as usize] != DRAW {
                    continue;
                }
                if result == CAT {
                    color[m][pc][CAT as usize] = CAT;
                    queue.push_back((m, pc, CAT, CAT));
                } else {
                    degree[m][pc][CAT as usize] -= 1;
                    if degree[m][pc][CAT as usize] == 0 {
                        color[m][pc][CAT as usize] = result;
                        queue.push_back((m, pc, CAT, result));
                    }
                }
            }
        } else {
            for &pm_i in &graph[m] {
                let pm = pm_i as usize;
                if color[pm][c][MOUSE as usize] != DRAW {
                    continue;
                }
                if result == MOUSE {
                    color[pm][c][MOUSE as usize] = MOUSE;
                    queue.push_back((pm, c, MOUSE, MOUSE));
                } else {
                    degree[pm][c][MOUSE as usize] -= 1;
                    if degree[pm][c][MOUSE as usize] == 0 {
                        color[pm][c][MOUSE as usize] = result;
                        queue.push_back((pm, c, MOUSE, result));
                    }
                }
            }
        }
    }

    color[1][2][MOUSE as usize]
}

fn main() {
    println!("{}", cat_mouse_game(vec![vec![2, 5], vec![3], vec![0, 4, 5], vec![1, 4, 5], vec![2, 3], vec![0, 2, 3]])); // 0
    println!("{}", cat_mouse_game(vec![vec![1, 3], vec![0], vec![3], vec![0, 2]])); // 1
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn examples() {
        assert_eq!(cat_mouse_game(vec![vec![2, 5], vec![3], vec![0, 4, 5], vec![1, 4, 5], vec![2, 3], vec![0, 2, 3]]), 0);
        assert_eq!(cat_mouse_game(vec![vec![1, 3], vec![0], vec![3], vec![0, 2]]), 1);
    }
}
